### ISSUE TITLE
Performance of current master is not enough for Booking.com

### ISSUE THREAD

**Issue by stepashka** [go-graphite/carbonapi#361]
The current HEAD of the master branch of carbonapi can not serve requests at a
high enough rate or quickly enough for Booking.com's needs. We're reasonably
happy with its correctness, as in that we think most of the concurrency bugs
and logic issues that we had seen have been resolved.

However, when we've tried rolling the code out into production, we've gotten
numerous complaints from our users that it's too slow to be useful, that most
of their requests return no data due to timeouts, and that automatic alerting
systems based on Graphite are no longer reliable.

Between May and July we made numerous attempts to upgrade
running into bugs and decreased performance every time we tried.
We followed up bisecting, fixing and making another attempt.

We could try to work further on these issues. We should also consider that the last
successful rollout of code from the master branches of carbonapi and
carbonzipper at Booking were in last March; we've been running version 0.9.2 of
carbonapi and 0.73 of carbonzipper since then.

There is some line of time that runs from that point in the past and into some
point in the future, when the master branch is again fit for use under our
production loads. We don't know, and can't know, how far away into the future
that point is, so the only way we see of moving forward on our end and serve
our customer's needs is to abandon the current direction.

We did this internally at the start of the month. We rewound the clock to
versions 0.11 of carbonapi and 0.74 of carbonzipper, merged the Git repos of
the projects, and cherry-picked commits from the master branch that were not
related to version 3 of the protobuf protocol or the logical merge of API and
Zipper functionalities.

We've rolled that code out, and will continue development in the same general
direction as current master. We hope to become as compatible as possible with
it, but can no longer justify the investment of work in its development.

Our focus in the near future will be on reliability of API and Zipper, and on
compatibility with the master branch. The general idea of our plans is to:

- Improve observability and error messages for debugging
- Enforce timeouts and cancellation of work that's no longer needed
- Merge the API and Zipper binaries
- Support version 3 of the protobuf protocol

As we said, much of this is in line with the objectives of the current master
branch, but we believe we have no option but to get to a similar place via an
alternate route.

The question we'd like to ask the community is what it would like us to do with
our changes? We don't expect them to take over the master branch, but we'd be
perfectly happy to keep them in a separate branch in this Github repo. That way
people could see what we're doing, we could keep exchanging patches on
non-infra related code, and so on.

We understand these are not good news. We're not very happy with this either,
but we have users to support, and don't feel we can do that as things stand
now.

**Comment by msaf1980** [comment]
We also have a performance problems on many concurrent queries. On low concurrency level - no problem, but on heavy load - very low rps (less than 8 times than commit 45279fe54c72c543aadf97372e590aa4248d1258). Some profiling results in attachment, but i can't  identify the problem.
[pprof.carbonapi.master.samples.cpu.001.pb.gz](https://github.com/go-graphite/carbonapi/files/2495241/pprof.carbonapi.master.samples.cpu.001.pb.gz)
[pprof.carbonapi.master.samples.cpu.001.pdf](https://github.com/go-graphite/carbonapi/files/2495242/pprof.carbonapi.master.samples.cpu.001.pdf)
[pprof.carbonapi.45279fe54c72c543aadf97372e590aa4248d1258.samples.cpu.001.pb.gz](https://github.com/go-graphite/carbonapi/files/2495243/pprof.carbonapi.45279fe54c72c543aadf97372e590aa4248d1258.samples.cpu.001.pb.gz)
[pprof.carbonapi.45279fe54c72c543aadf97372e590aa4248d1258.samples.cpu.001.pdf](https://github.com/go-graphite/carbonapi/files/2495244/pprof.carbonapi.45279fe54c72c543aadf97372e590aa4248d1258.samples.cpu.001.pdf)

**Comment by azhiltsov** [comment]
@msaf1980 probably you are facing the same problem as we did and all what I can recommend currently is to try our fork which we currently developing and running in production. Here is the link: https://github.com/bookingcom/carbonapi/

@Civil feel free to update the Readme

**Comment by Civil** [comment]
@msaf1980 I think I found the issue after discussing that behaviour in #392 

It seems that the reason of low performance is wrong usage of limiter logic, introduced in one of the earlier commits that I somehow overlooked during the review.

It resulted in unexpected increase of waiting time even if concurrency was set to high values (as query was waiting on a wrong limiter). This should be fixed in the current master.

Also in this case pprof would be not very useful to debug, but it should be caught by mutex profiler.

I guess after I'll receive confirmation in #392 this issue can be closed as it will be no longer possible to validate this statement as it doesn't seem that booking.com will be willing to try migrate back to this carbonapi.

**Comment by Civil** [comment]
Taking into account my last comment, I feel that there is nothing left here to discuss.

Feel free to reopen if I'm mistaken.

**Comment by Civil** [comment]
@msaf1980 in current master, I've raised the default 'concurency' limiter to 1000 (was 20). It's likely that out-of-box performance should be much better now (if you see no CPU usage and very low performance - likely you hit this limiter)
